util/tracecpu/collector.go

```diff
@@ -0,0 +1,391 @@
+// Copyright 2021 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracecpu
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/pingcap/log"
+	"github.com/pingcap/tipb/go-tipb"
+	"github.com/wangjohn/quickselect"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+)
+
+// TopSQLRecord represents a single record of how much cpu time a sql plan consumes in one second.
+//
+// PlanDigest can be empty, because:
+// 1. some sql statements has no plan, like `COMMIT`
+// 2. when a sql statement is being compiled, there's no plan yet
+type TopSQLRecord struct {
+	SQLDigest  []byte
+	PlanDigest []byte
+	CPUTimeMs  uint32
+}
+
+type planBinaryDecodeFunc func(string) (string, error)
+
+// topSQLDataPoints represents the cumulative SQL plan CPU time in current minute window
+type topSQLDataPoints struct {
+	SQLDigest      []byte
+	PlanDigest     []byte
+	CPUTimeMsList  []uint32
+	TimestampList  []uint64
+	CPUTimeMsTotal uint64
+}
+
+// cpuTimeSort is used to sort TopSQL records by tocal CPU time
+type cpuTimeSort struct {
+	Key            string
+	SQLDigest      []byte
+	PlanDigest     []byte
+	CPUTimeMsTotal uint64
+}
+type cpuTimeSortSlice []cpuTimeSort
+
+func (t cpuTimeSortSlice) Len() int {
+	return len(t)
+}
+
+// We need find the kth largest value, so here should use >
+func (t cpuTimeSortSlice) Less(i, j int) bool {
+	return t[i].CPUTimeMsTotal > t[j].CPUTimeMsTotal
+}
+func (t cpuTimeSortSlice) Swap(i, j int) {
+	t[i], t[j] = t[j], t[i]
+}
+
+type planRegisterJob struct {
+	planDigest     string
+	normalizedPlan string
+}
+
+// TopSQLCollector is called periodically to collect TopSQL resource usage metrics
+type TopSQLCollector struct {
+	// // calling this can take a while, so should not block critical paths
+	// planBinaryDecoder planBinaryDecodeFunc
+	mu sync.RWMutex
```

Comment:
For performance, better use different `mutex` for `normalizedPlanMap` and `normalizedPlanMap`.
Or use `sync.Map` maybe a better choise?

Reply:
Will use `sync.Map`